--- gtk+-2.24.0/gdk/quartz/gdkcolor-quartz.c.orig	2012-08-13 22:23:03.000000000 +0400
+++ gtk+-2.24.0/gdk/quartz/gdkcolor-quartz.c	2012-08-13 22:24:50.000000000 +0400
@@ -237,7 +237,7 @@
         else
           components[3] = 1.0;
 
-        colorspace = CGColorSpaceCreateWithName (kCGColorSpaceGenericRGB);
+        colorspace = CGColorSpaceCreateDeviceRGB();
         color = CGColorCreate (colorspace, components);
         CGColorSpaceRelease (colorspace);
         break;
