diff --git a/gdk/quartz/gdkkeys-quartz.c b/gdk/quartz/gdkkeys-quartz.c
index e4c1636..1cb7a3a 100644
--- a/gdk/quartz/gdkkeys-quartz.c
+++ b/gdk/quartz/gdkkeys-quartz.c
@@ -155,7 +155,7 @@ const static struct {
   { 67, GDK_asterisk, GDK_KP_Multiply },
   { 69, GDK_plus, GDK_KP_Add },
   { 75, GDK_slash, GDK_KP_Divide },
-  { 76, 0x01000003, GDK_KP_Enter },
+  { 76, GDK_Return, GDK_KP_Enter },
   { 78, GDK_minus, GDK_KP_Subtract },
   { 81, GDK_equal, GDK_KP_Equal },
   { 82, GDK_0, GDK_KP_0 },
