--- a/gdk/quartz/gdkglobals-quartz.c
+++ b/gdk/quartz/gdkglobals-quartz.c
@@ -30,10 +30,12 @@ GdkWindow *_gdk_root = NULL;
 GdkOSXVersion
 gdk_quartz_osx_version (void)
 {
-  gint32 minor;
-  OSErr err = Gestalt (gestaltSystemVersionMinor, (SInt32*)&minor);
-  g_return_val_if_fail (err == noErr, GDK_OSX_UNSUPPORTED);
-
+  static gint32 minor = GDK_OSX_UNSUPPORTED;
+  if (minor == GDK_OSX_UNSUPPORTED)
+    {
+       OSErr err = Gestalt (gestaltSystemVersionMinor, (SInt32*)&minor);
+       g_return_val_if_fail (err == noErr, GDK_OSX_UNSUPPORTED);
+    }
   if (minor < GDK_OSX_MIN)
     return GDK_OSX_UNSUPPORTED;
   else if (minor > GDK_OSX_CURRENT)
