--- a/gdk/quartz/gdkgc-quartz.c
+++ b/gdk/quartz/gdkgc-quartz.c
@@ -289,7 +289,10 @@ _gdk_windowing_gc_set_clip_region (GdkGC           *gc,
       private->have_clip_mask = FALSE;
     }
 
-  private->have_clip_region = region != NULL;
+  if (region == NULL || gdk_region_empty (region))
+     private->have_clip_region = FALSE;
+  else
+     private->have_clip_region = TRUE;
 
   if (reset_origin)
     {
